File: src/WellTool.Json/hutool-json/src/test/java/cn/hutool/json/JSONPathTest.java
package cn.hutool.json;

import static org.junit.jupiter.api.Assertions.*;

import cn.hutool.core.lang.TypeReference;
import org.junit.jupiter.api.Test;

import java.util.List;

/**
 * JSON路径单元测试
 *
 * @author looly
 *
 */
public class JSONPathTest {

	@Test
	public void getByPathTest() {
		String json = "[{\"id\":\"1\",\"name\":\"xingming\"},{\"id\":\"2\",\"name\":\"mingzi\"}]";
		Object value = JSONUtil.parseArray(json).getByPath("[0].name");
		assertEquals("xingming", value);
		value = JSONUtil.parseArray(json).getByPath("[1].name");
		assertEquals("mingzi", value);
	}

	@Test
	public void getByPathTest2(){
		String str = "{'accountId':111}";
		JSON json = JSONUtil.parse(str);
		Long accountId = JSONUtil.getByPath(json, "$.accountId", 0L);
		assertEquals(111L, accountId.longValue());
	}

	@Test
	public void getByPathTest3(){
		String str = "[{'accountId':1},{'accountId':2},{'accountId':3}]";
		JSON json = JSONUtil.parse(str);
		// 返回指定泛型的对象 List<Long>
		List<Long> accountIds = json.getByPath("$.accountId", new TypeReference<List<Long>>() {
		});
		assertNotNull(accountIds);
		assertArrayEquals(new Long[]{1L, 2L, 3L}, accountIds.toArray());

		str = "{\"accountInfos\": [{\"accountId\":1},{\"accountId\":2},{\"accountId\":3}]}";
		json = JSONUtil.parse(str);
		// 返回指定泛型的对象 List<Long>
		accountIds = json.getByPath("$.accountInfos.accountId", new TypeReference<List<Long>>() {
		});
		assertNotNull(accountIds);
		assertArrayEquals(new Long[]{1L, 2L, 3L}, accountIds.toArray());
	}

	@Test
	public void getByPathWithWildcardTest() {
		// 测试通配符 [*] 语法
		JSONObject root = new JSONObject()
			.put("actionMessage", new JSONObject()
				.put("alertResults", new JSONArray())
				.put("decodeFeas", new JSONArray()
					.put(new JSONObject()
						.put("body", new JSONObject()
							.put("lats", new JSONArray()
								.put(new JSONObject().put("begin", 4260).put("text", "呵呵"))
								.put(new JSONObject().put("begin", 4260).put("text", "你好 "))
							)
						)
					)
				)
			);

		Object byPath = JSONUtil.getByPath(root, "$.actionMessage.decodeFeas[0].body.lats[*].text");
		assertNotNull(byPath);
		assertEquals("[呵呵, 你好 ]", byPath.toString());
	}
}

File: src/WellTool.Json.Tests/JSONPathTest.cs
using System;
using System.Collections.Generic;
using Xunit;

namespace WellTool.Json.Tests
{
    /// <summary>
    /// JSONPath 测试
    /// </summary>
    public class JSONPathTest
    {
        [Fact]
        public void GetByPathTest()
        {
            var json = "[{\"id\":\"1\",\"name\":\"xingming\"},{\"id\":\"2\",\"name\":\"mingzi\"}]";
            var array = JSONUtil.ParseArray(json);
            var value = array.GetByPath("[0].name");
            Assert.Equal("xingming", value);
            value = array.GetByPath("[1].name");
            Assert.Equal("mingzi", value);
        }

        [Fact]
        public void GetByPathTest2()
        {
            var str = "{'accountId':111}";
            var json = JSONUtil.Parse(str);
            var accountId = JSONUtil.GetByPath(json, "$.accountId", 0L);
            Assert.Equal(111L, accountId);
        }

        [Fact]
        public void GetByPathTest3()
        {
            var str = "[{'accountId':1},{'accountId':2},{'accountId':3}]";
            var json = JSONUtil.Parse(str);
            var accountIds = json.GetByPath<List<long>>("$.accountId");
            Assert.NotNull(accountIds);
            Assert.Equal(3, accountIds.Count);
            Assert.Equal(1L, accountIds[0]);
            Assert.Equal(2L, accountIds[1]);
            Assert.Equal(3L, accountIds[2]);
        }

        [Fact]
        public void GetByPathWithWildcardTest()
        {
            var root = new JSONObject()
                .SetOpt("actionMessage", new JSONObject()
                    .SetOpt("alertResults", new JSONArray())
                    .SetOpt("decodeFeas", new JSONArray()
                        .SetOpt(new JSONObject()
                            .SetOpt("body", new JSONObject()
                                .SetOpt("lats", new JSONArray()
                                    .SetOpt(new JSONObject().SetOpt("begin", 4260).SetOpt("text", "呵呵"))
                                    .SetOpt(new JSONObject().SetOpt("begin", 4260).SetOpt("text", "你好 "))
                                )
                            )
                        )
                    )
                );

            var byPath = JSONUtil.GetByPath(root, "$.actionMessage.decodeFeas[0].body.lats[*].text");
            Assert.NotNull(byPath);
            Assert.Equal("[\"呵呵\",\"你好 \"]", byPath.ToString());
        }
    }
}
